extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,204 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fs::File,
+    io::Read,
+    path::PathBuf,
+};
+
+use clap::Parser;
+use fingerprint::Combined;
+use getset::Getters;
+use serde::{Deserialize, Serialize};
+use stable_eyre::{eyre::Context, Report};
+use tar::{Archive, Entry};
+use thiserror::Error;
+use tracing::{debug, info, info_span, warn};
+use typed_builder::TypedBuilder;
+
+/// Tar errors are just [`std::io::Error`].
+/// This macro is a bit cleaner to wrap them and adds a little context.
+macro_rules! wrap_tar_err {
+    ($tar_operation_str:expr, $operation:expr) => {
+        $operation.map_err(|e| Error::ReadTar($tar_operation_str.to_string(), e))
+    };
+}
+
+#[derive(Debug, Parser, Getters)]
+#[getset(get = "pub")]
+#[clap(version)]
+pub struct Subcommand {
+    /// The tar file image to search and fingerprint JARs in.
+    image_tar_file: PathBuf,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
+struct DiscoveredJar {
+    path: PathBuf,
+    fingerprint: Combined,
+}
+
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+struct OciManifest {
+    #[serde(rename = "Layers")]
+    layers: Vec<PathBuf>,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, TypedBuilder)]
+struct JarAnalysis {
+    // A vector of the Jars discovered during an analysis.
+    discovered_jars: HashMap<PathBuf, Vec<DiscoveredJar>>,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Encountered when we can't open the requisite file.
+    #[error("Open file {0}")]
+    OpenFile(std::io::Error),
+
+    /// Encountered when an archive operation fails.
+    #[error("Tar operation '{0}': {1}")]
+    ReadTar(String, std::io::Error),
+
+    /// Encountered when a manifest file is found but cannot be parsed.
+    #[error("Parsing manifest file {0} {1}")]
+    ParseManifest(PathBuf, serde_json::Error),
+
+    #[error("Reading buffer {0}")]
+    ReadBuffer(std::io::Error),
+
+    #[error("Generating fingerprint {0}")]
+    GenerateFingerprint(fingerprint::Error),
+}
```

Comment:
We're using `stable_eyre` elsewhere, why do we even want an error enum here vs ad-hoc `.context()`?

Reply:
I think I started down the road of using an enum and then kept going. I think ad-hoc `.context()` is fine since there's nowhere I actually inspect the results. I will make this change.

Reply:
I think I've resolved this and other feedback. Going to wait to merge until more is done on the second PR. Thanks for having a look!